ticket_booking_app/src/main/java/com/mordor/service/ReservationServiceImpl.java

```diff
@@ -0,0 +1,245 @@
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.annotation.PostConstruct;
+import org.apache.http.client.utils.URIBuilder;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+import org.springframework.stereotype.Service;
+import com.mordor.Utils.ResourceNotFoundException;
+import com.mordor.Utils.RunnableExpirationTask;
+import com.mordor.Utils.WrongReservationException;
+import com.mordor.dao.ReservationDAO;
+import com.mordor.dao.SeatReservationDAO;
+import com.mordor.model.ReservationConfirmation;
+import com.mordor.model.converter.ConverterDTO;
+import com.mordor.model.dto.ReservationConfirmationDTO;
+import com.mordor.model.dto.ReservationDTO;
+import com.mordor.model.dto.TicketDTO;
+import com.mordor.model.enitity.MovieScreening;
+import com.mordor.model.enitity.Reservation;
+import com.mordor.model.enitity.Seat;
+import com.mordor.model.enitity.SeatReservation;
+
+@Service
+public class ReservationServiceImpl implements ReservationService{
+	
+	@Autowired
+	ReservationDAO reservationDAO;
+	
+	@Autowired
+	SeatReservationDAO seatReservationDAO;        
+	
+	@Autowired
+	SeatReservationService seatReservationService;
+	
+	@Autowired
+	SeatService seatService;
+	
+	@Autowired
+	MovieScreeningService movieScreeningService;
+	
+	@Autowired
+	ThreadPoolTaskScheduler threadPoolTaskScheduler;
+	
+	@Autowired
+	private ConverterDTO<ReservationConfirmationDTO, ReservationConfirmation> reservationConfirmationConverter;
+	
+	@Autowired
+	private ConverterDTO<ReservationDTO, Reservation> reservationConverter;
+	
+	@Autowired
+	@Qualifier("ticketConverter")
+	private ConverterDTO<TicketDTO, SeatReservation> ticketConverter;
+	
+	@PostConstruct
+	private void init() {
+		threadPoolTaskScheduler.setPoolSize(10);
+	}
+	
+	
+	@Override
+	public ReservationConfirmationDTO saveReservation(ReservationDTO reservationDTO) throws MalformedURLException, ResourceNotFoundException, WrongReservationException {
+		Reservation reservation = reservationConverter.convertToEntity(reservationDTO);
+		
+		if(checkLeftOverSeat(reservationDTO.getTickets())) {
+			throw new WrongReservationException("Left over seats");
+		}
+		
+		Instant timeofReservation = Instant.now(Clock.offset(Clock.systemUTC(), Duration.ofHours(2)));
+		
+		reservation.setExpirationTime(ReservationService.getExpirationTime(timeofReservation));
+		
+		List<SeatReservation> seatReservations = reservationDTO.getTickets().stream()
+				.map(ticketDTO -> {
+					try {
+						return ticketConverter.convertToEntity(ticketDTO);
+					} catch (ResourceNotFoundException e) {
+						e.printStackTrace();

```

Comment:
To nie jest najlepszy sposób radzenia sobie z wyjątkami. Tu poleci exception, a metoda wołająca nic o tym nie będzie wiedzieć.